File: Medium/2025-10-07-1488-Avoid-Flood-in-The-City/solution.js
var avoidFlood = function(rains) {
    const n = rains.length;
    const ans = new Array(n).fill(-1);
    const fullLakes = new Set();
    const dryDays = [];
    
    for (let i = 0; i < n; i++) {
        if (rains[i] > 0) {
            if (fullLakes.has(rains[i])) {
                return [];
            }
            fullLakes.add(rains[i]);
        } else {
            dryDays.push(i);
        }
    }
    
    for (let i = 0; i < n; i++) {
        if (rains[i] > 0) {
            ans[i] = -1;
            const lake = rains[i];
            fullLakes.delete(lake);
        } else {
            if (fullLakes.size > 0) {
                const lakeToDry = fullLakes.values().next().value;
                ans[i] = lakeToDry;
                fullLakes.delete(lakeToDry);
            } else if (dryDays.length > 0) {
                ans[i] = 1; // Dry any lake (1 is arbitrary)
            }
        }
    }
    
    return ans;
};

File: Medium/2025-10-07-1488-Avoid-Flood-in-The-City/solution.java
import java.util.*;

class Solution {
    public int[] avoidFlood(int[] rains) {
        int n = rains.length;
        int[] ans = new int[n];
        Arrays.fill(ans, -1);
        
        Set<Integer> fullLakes = new HashSet<>();
        Queue<Integer> dryDays = new PriorityQueue<>();
        
        for (int i = 0; i < n; i++) {
            if (rains[i] > 0) {
                int lake = rains[i];
                if (fullLakes.contains(lake)) {
                    return new int[0]; // Flood occurs
                }
                fullLakes.add(lake);
                ans[i] = -1; // It rained, no drying
                dryDays.offer(i); // Store the day we can dry this lake
            } else {
                ans[i] = 1; // Default to drying lake 1
            }
            
            // Dry lakes on days where it rained before
            while (!dryDays.isEmpty() && !fullLakes.isEmpty()) {
                int dryDay = dryDays.poll();
                if (ans[dryDay] == 1) {
                    int lakeToDry = fullLakes.iterator().next();
                    fullLakes.remove(lakeToDry);
                    ans[dryDay] = lakeToDry; // Dry this lake
                }
            }
        }
        
        return ans;
    }
}